Add EmptyObject

jonathanjouty opened the pull request:
I also added a WIP commit (034395640b1d2fa7f83b6c2acf3b771ec06e2bcb) for the version bump and `CHANGELOG.md` update.
I can remove it if you prefer.

jonathanjouty committed:
Add EmptyObject

jonathanjouty committed:
WIP: Bump version

TODO: Finalise release date

deviant-logic commented:
Hi Jonathan, sorry for taking so long to get back to you.  This is a great addition, which I'm absolutely happy to merge as-is, but before Thanksgiving week distracted me, I was thinking about whether it would make sense to have `newtype EmptyObject = EmptyObject { emptyObject :: () }`?  It looks a little funny, but it follows the `emptyObject <$> decode "{}"` pattern, though I suppose it makes encoding a little more awkward (`encode $ EmptyObject ()`).  Anyway, I'm not sure which variant has better ergonomics, but I wanted to suggest it and see what you thought.  I'm happy to merge either variant.  What do you think?

jonathanjouty commented:
@deviant-logic No problem.
For the suggestion: I will try this out in our code this week and get back to you as to how it affects things.

jonathanjouty commented:
This is how I'm using it currently, so it makes little difference:
`data EmptyObject`
```hs
mMyVal = MyNullaryCons <$ (parseMaybe parseJSON v :: Maybe EmptyObject)
```
vs `newtype EmptyObject`
```hs
mMyVal = MyNullaryCons <$ emptyObject <$> parseMaybe parseJSON v
```

I imagine in most situations you will want to consume an `EmptyObject` rather than encode one, so let's go with your suggestion.

jonathanjouty committed:
Use a newtype for EmptyObject for better ergonomics

As suggested here https://github.com/deviant-logic/aeson-filthy/pull/3#issuecomment-560166314
